### ISSUE TITLE
Introduce env variable for cli switch `--project-directory`

### ISSUE THREAD

**Issue by thomas-mc-work** [docker/compose#12200]
### Description

I know there is `COMPOSE_FILE` to specify the path to a compose file. But the problem is that this can be either of `docker-compose.yml` or `docker-compose.override.yml`. It's is choosen based on a rule when docker compose is launched without options. I want this behaviour using an env variable too, like:

    COMPOSE_PROJECT_DIRECTORY

**Comment by thomas-mc-work** [comment]
Here is the result of some commands:
(The service comprises of both `docker-compose.yml` and `docker-compose.overrride.yml`)

- `docker compose -f /my-service/docker-compose.yml down` is working as expected
- `docker compose -f /my-service/docker-compose.yml up` isn't respecting `docker-compose.overrride.yml`
- `docker compose -f /my-service/docker-compose.overrride.yml up` is complaining because about missing definitions in the file (which is correct when it is seen standalone)
- `docker compose -f /my-service/docker-compose.yml config` same as `… up`
- `docker compose -f /my-service/docker-compose.overrride.yml config` same as `… up`

**Comment by ndeloof** [comment]
why not use `docker compose --project-directory /my-service <command>` ?

**Comment by thomas-mc-work** [comment]
I don't want to clutter the commands / they shall stay portable. So I can include them in script which aren't aware of the fact of the actual project path. Isn't that the same reason to have the `COMPOSE_FILE` env variable?

**Comment by ndeloof** [comment]
yes it is (while we shut into our own foot with this as this triggered a huge amount of issues :P)
I'm fine we introduce support for `COMPOSE_PROJECT_DIRECTORY` as an alternative for flag `--project-directory`, just wanted to confirm this covers your needs.

**Comment by thomas-mc-work** [comment]
> while we shut into our own foot with this as this triggered a huge amount of issues

I hope you didn't had to introduce workarounds for it and it was just cleaning up internal structures. You also have to judge whether my request is valid in terms of the introduction of complexity and maintenance burden. But I'd of course be happy if it could happen :-)

**Comment by ndeloof** [comment]
Unfortunately introduction of  `COMPOSE_FILE` support and interractionwith `.env` file created a huge amount of issues and regressions, had to fix back and forth in a few places, so I'm reluctant with any additional move taking place in this area.

Your use case is easily supported by a plain `cd xxx ; docker compose up` or use of `--project-directory $COMPOSE_PROJECT_DIRECTORY` in your script, so I prefer to say "no" _vs_ risk for more regressions being introduced :)

**Comment by thomas-mc-work** [comment]
I hope you don't mind me to reuse this issue for a final question:

> Unfortunately introduction of COMPOSE_FILE support and interractionwith .env file created a huge amount of issues and regressions, had to fix back and forth in a few places, so I'm reluctant with any additional move taking place in this area.

Without knowing the source code at all (but having sound experiences with that issue in other languages) I wonder if it really can cause any impacts on the processing part. What I think is that this switch should be applied at the very beginning of the process, same as the CLI switch does. Only after that the logic should follow, independent of where the variable value is from.

There is an outdated pull request targteing the same issue from the times when the code was based on Python: 

https://github.com/docker/compose/pull/5064

I know this is not applicable anymore. But I wonder if the core principle has changed.

Thank you for taking the time to share your opinion withme!

**Comment by ndeloof** [comment]
> I wonder if the core principle has changed.

it has. Basically, ability to set COMPOSE_* variable has been introduced in .env file, ... which can be set by --env-file _or_ COMPOSE_ENV_FILE, so a chicken-egg issue has been introduced and the codebase hardly can manage this without some issues.
